### ISSUE TITLE
[Question] How to execute JavaScript code on each page load event?

### ISSUE THREAD

**Issue by mareru** [microsoft/playwright#19030]
Hello, is there a way to execute JavaScript code on each page load event for every page in the app in Playwright?   
So basically every time page is refreshed (loads again) I would like to execute my JavaScript code.  
I tried something like this - add it to my auto fixture, but Typescript complains that "Promise returned in function argument where a void return was expected."  
```Typescript
zoomTo100Percent: [
    async ({ page }, use) => {
      page.on('load', async (page) => {
        await page.evaluate(() => {
          const body = document.querySelector('body');
          if (body) {
            // Removes the class only if it exists on the body tag
            body.classList.remove('zoom-out');
          } else {
            throw Error(ErrorMessage.BODY_NOT_FOUND);
          }
        });
      })
      await use();
    },
    { auto: true },
  ],
```

I also tried this but didn't help as well my zoom-out class is not removed from the page:
```Typescript
zoomTo100Percent: [
    async ({ page }, use) => {
      await page.addInitScript(async () => {
        await page.evaluate(() => {
          const body = document.querySelector('body');
          if (body) {
            // Removes the class only if it exists on the body tag
            body.classList.remove('zoom-out');
          } else {
            throw Error(ErrorMessage.BODY_NOT_FOUND);
          }
        });
      });
      await use();
    },
    { auto: true },
  ],
```
I also tried something like this but it didn't work as well:  
```Typescript
  page: async ({ page }, use) => {
   console.log('I am in page fixture override');
    await page.addInitScript(async () => {
      console.log('I am in the addInitScript');
      await page.evaluate(() => {
        const body = document.querySelector('body');
        if (body) {
          // Removes the class only if it exists on the body tag
          body.classList.remove('zoom-out');
          console.log('Removing the zoom-out class');
          console.log('Class exists: ', body.classList.contains('zoom-out'));
        } else {
          throw Error(ErrorMessage.BODY_NOT_FOUND);
        }
      });
    });
    await use(page);
  },
```
The zoom-out class is directly on the body, but it is a part of an HTML if statement. Could it be that it is appended only after the scripts are added, so when the code above runs it makes no effect since the class is added only after the scripts were run and the above code is evaluated before the scripts were run?

It seems like the function in addInitScript was never executed.  
I got this logged to the console: `I am in page fixture override` but not the rest of the console logs in the addInitScript 
Why is that so? Can you please help?  
Thanks!

**Comment by dgozman** [comment]
@mareru You are on the right track! `page.addInitScript` already executes in the page, so you don't need `page.evaluate`. Also, `addInitScript` is called before `body` is inserted, so you need to wait `DOMContentLoaded`.

```ts
  page: async ({ page }, use) => {
    await page.addInitScript(() => {
     // Make sure body has loaded.
      window.addEventListener('DOMContentLoaded', () => {
        new MutationObserver(() => {
          // Whenever "zoom-out" is added, remove it.
          if (document.body.classList.contains('zoom-out'))
            document.body.classList.remove('zoom-out');
        }).observe(document.body, { attributes: true });
      });
    });
    await use(page);
  },
```

**Comment by mareru** [comment]
@dgozman 
Thank you a ton! This worked as a charm. 
I can express my gratitude in words I would like to come to CA and hug you :) 
This is such a good community - I always get so much response and commitment from people.  
I am so happy. 
Thank you once again! 👍

**Comment by ivanlgarcia28** [comment]
@mareru hi! Thank you for answer me and sorry for the missinformation by my side. This is the first time I interact here and also the first time I interact with iFrames, so I concluded that was my problem.
I've tried a lot of possible solutions but I found one that works for me:

`async populateBody(){
    const frame= this.page.frameLocator("#iframe-content");
    await frame.locator("#paragraph").type("Lorem Ipsum");
}`

The page code is something like this:
`<iframe>
    <html>
        <body id="iframe-content">
            <p id="paragraph">

            </p>
        </body>
    </html>
</iframe>`

So, my conclusion is even I tried .fill(), .innerText and evaluate() method, I wasn't realize that I'm working with iFrames, and .fill() method only works on input, text-area elements.

Again, thank you so much for your attention and I hope If someone has any similar issues like me, this can help.
